### PR TITLE
 Port to Js_of_ocaml

### PR DESCRIPTION
**Pull request by mnxn**
I have completed the port from BuckleScript to Js_of_ocaml. From my testing, all of the features continue to be functional under both opam and esy sandboxes. 

All bucklescript-specific code and configuration files have been removed. The VSCode npm dependency has been bumped to 1.49 because the VSCode bindings are written around that version.

Although the size of the generated JS has increased to about 16MB with Js_of_ocaml, the installation vsix package has only increased to 3.5MB from BuckleScript's 1.0MB.

A few opam dependencies have been introduced:

- [js_of_ocaml](https://opam.ocaml.org/packages/js_of_ocaml/) and [gen_js_api](https://opam.ocaml.org/packages/gen_js_api/) for creating JS bindings.

- [core_kernel](https://opam.ocaml.org/packages/core_kernel/) for a few instances where the OCaml stdlib does not provide certain functionality. We could also opt to use Core_kernel as a complete stdlib replacement.

- [promise_jsoo](https://opam.ocaml.org/packages/promise_jsoo/) for gen_js_api-compatible bindings to JS Promises. This library includes variations of the functions that were previously in `Promise.ml`.

- [jsonoo](https://opam.ocaml.org/packages/jsonoo/) for gen_js_api-compatible JSON type and convertors. This library replaces the functionality of bs-json.

### CODE REVIEW

**Comment by mnxn** [comment]
It took me a while, but I finally figured out the CI. 

We will have to continue to use OCaml 4.06 in the CI for running the bsb and opam tests, while OCaml 4.10 will be used for building the extension and running esy tests. 

- OCaml 4.06 can't be used to build the extension itself because some of the dependencies need at least 4.08. 
- OCaml 4.11 can't be used to build the extension because the latest published ReasonML package doesn't support it yet.
- OCaml 4.10 can't be used in the CI because BuckleScript requires 4.06.

The JS file is built by `esy build` in the CI and `yarn package` no longer calls the build command. It's a bit confusing having this mix of build systems in a single project, so there's definitely room for improvement. I think it would be useful to move the remaining commands from the `package.json` to the Makefile to make it simpler.

**Comment by smorimoto** [comment]
Amazing work! I just skimmed it on the move but it looks good overall. What does everyone else think?

**Review by smorimoto — APPROVED** [review]

File: esy.json [hunk 1]
```diff
@@ -1,8 +1,20 @@
 {
+  "esy": {
+    "build": "make build",
```
Comment (rgrinberg) [inline esy.json:3]: Can we call dune directly?

File: esy.json [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
+    "buildsInSource": "_build"
```
Comment (rgrinberg) [inline esy.json:4]: Why do we need this line?

File: src/DuneTaskProvider.ml [hunk 3]
```diff
@@ -4,38 +4,42 @@ type t = Disposable.t option ref
...
 let source = task_type
 
-let problemMatchers = [| "$ocamlc" |]
+let problem_matchers = [ "$ocamlc" ]
 
 (* the ocamlc matcher is not able to parse ocaml compiler errors unless they
    follow the short style. *)
-let env = Js.Dict.fromList [ ("OCAML_ERROR_STYLE", "short") ]
+let env =
+  Core_kernel.Hashtbl.of_alist_exn
```
Comment (rgrinberg) [inline src/DuneTaskProvider.ml:16]: Can this be a map? I don't think it's right to modify this.

File: dune-project [hunk 4]
```diff
@@ -0,0 +1 @@
+(lang dune 2.7)
```
Comment (rgrinberg) [inline dune-project:1]: Let's add a `(name ..)` field here.

File: esy.json (hunk 2 above)
Comment (mnxn) [reply esy.json:4]: That line is so the esy build location is consistent with normal dune. The `main` is specified to be `_build/default/src/extension.bc.js` in the package.json. If that line wasn't there, it would be built under the `_esy` folder.

File: src/Cmd.ml [hunk 5]
```diff
@@ -29,12 +29,12 @@ let candidateFns =
 
 let which path fn =
   let candidates = candidateFns fn in
-  Js.String.split envSep path
-  |> Promise.Array.findMap (fun p ->
+  Core_kernel.String.split ~on:envSep path
```
Comment (rgrinberg) [inline src/Cmd.ml:32]: You should `open Core_kernel` in import. will make this a bit simpler to read.

File: src/Cmd.ml [hunk 6]
```diff
@@ -55,39 +55,48 @@ let check t =
...
+      :: ("args", list string args)
       :: message
-    | Shell commandLine -> ("shell", Log.field commandLine) :: message
+    | Shell commandLine -> ("shell", string commandLine) :: message
   in
   logJson "external command" message
 
 let output ?stdin (t : t) =
-  let open Promise.O in
+  let open Promise.Syntax in
```
Comment (rgrinberg) [inline src/Cmd.ml:93]: Why was this renamed?

File: esy.json (hunk 1 above)
Comment (mnxn) [reply esy.json:3]: I suppose we could, but that would be duplicating `dune build src/extension.bc.js` in two places.

Is there was a way for `dune build` to build that JS file automatically?

File: src/Output.ml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 let languageServerOutputChannel =
-  lazy (Vscode.Window.createOutputChannel "OCaml Language Server")
+  lazy (Vscode.Window.create_output_channel ~name:"OCaml Language Server")
```
Comment (rgrinberg) [inline src/Output.ml:2]: Why was this renamed? When creating bindings we should match original names whenever possible. Makes our bindings a lot easier to use.

File: src/Toolchain.ml [hunk 8]
```diff
@@ -240,25 +241,25 @@ let selectPackageManager (choices : Candidate.t list) =
         (quickPick, pm))
       choices
   in
-  let options =
-    Window.QuickPickOptions.make ~canPickMany:false ~placeHolder ()
-  in
-  Window.showQuickPickItems choices options
+  let options = QuickPickOptions.create ~can_pick_many:false ~place_holder () in
+  Window.show_quick_pick_items ~choices ~options ()
```
Comment (rgrinberg) [inline src/Toolchain.ml:245]: Ditto. The old name made it obvious how to look up the docs.

File: src/bindings/dune [hunk 9]
```diff
@@ -0,0 +1,27 @@
+(library
+ (name interop)
+ (modules interop)
+ (preprocess
+  (pps gen_js_api.ppx))
+ (libraries core_kernel js_of_ocaml gen_js_api)
+ (modes byte))
```
Comment (rgrinberg) [inline src/bindings/dune:7]: Could we put these libraries in separate subdirs?

File: src/bindings/dune [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ (preprocess
+  (pps gen_js_api.ppx))
+ (js_of_ocaml
+  (javascript_files node_stub.js))
+ (libraries interop core_kernel promise_jsoo jsonoo gen_js_api)
+ (modes byte))
+
+(library
+ (name vscode)
+ (modules vscode vscode__core vscode__languageclient)
```
Comment (rgrinberg) [inline src/bindings/dune:21]: Why are you manually using the `__` convention here? this is just a convention chosen by dune, we shouldn't use it.

File: src/dune [hunk 11]
```diff
@@ -0,0 +1,5 @@
+(executable
+ (name extension)
```
Comment (rgrinberg) [inline src/dune:2]: I'd rename this executable to `vscode-ocaml-platform`. The name `extension` is a bit too generic.

File: src/path.ml [hunk 12]
```diff
@@ -2,19 +2,19 @@ type t = string
 
 let ofString s = s
 
-let isAbsolute t = Node.Path.isAbsolute t
+let isAbsolute t = not (Filename.is_relative t)
```
Comment (rgrinberg) [inline src/path.ml:5]: We can now use the standard naming convention here.

File: esy.json (hunk 1 above)
Comment (rgrinberg) [reply esy.json:3]: We could introduce an alias. E.g. `dune build @vscode` or even just `$ dune build @all`.

File: src/Cmd.ml (hunk 6 above)
Comment (rgrinberg) [reply src/Cmd.ml:93]: The `O` convention is what we use in dune to mean the same thing. I like it a little more than `Syntax` which is quite verbose. In any case, I don't think it matters here. I don't mind following core's convention.

File: src/opam.ml [hunk 13]
```diff
@@ -21,24 +21,25 @@ let binary = Path.ofString "opam"
...
   | Ok cmd -> Some cmd
 
 let parseSwitchList out =
   let lines = String.split_on_char '\n' out in
   let result =
-    Belt.List.keepMap lines (function
-      | "" -> None
-      | s -> Some (Switch.make s))
+    lines
+    |> List.filter_map (function
```
Comment (rgrinberg) [inline src/opam.ml:33]: This module should also use Core_kernel

Comment (mnxn) [reply src/opam.ml:33]: So is it all right to completely transition to using Core_kernel?
I didn't do that when I made the PR because I didn't want to make too many unnecessary changes.

**Comment by rgrinberg** [comment]
CI needs to be updated to 4.10, I think?

File: src/opam.ml (hunk 13 above)
Comment (rgrinberg) [reply src/opam.ml:33]: Okay, it can  be done later. I saw other files ported to Core_kernel, so I thought you were doing it here too.

**Comment by mnxn** [comment]
> CI needs to be updated to 4.10, I think?

I mentioned in [this comment](https://github.com/ocamllabs/vscode-ocaml-platform/pull/385#issuecomment-703183258) that we can't do that as long as we're still testing for BuckleScript, which requires the 4.06 version of the tools.

**Comment by rgrinberg** [comment]
> I mentioned in this comment that we can't do that as long as we're still testing for BuckleScript, which requires the 4.06 version of the tools.

Oops sorry. Missed that.

I think we can just drop the bucklescript tests. We don't officially support bucklescript anymore. We'll still fix bucklescript bugs in the short term (~6 month), but by then, I hope that most people would transition to the official rescript extension.

**Comment by rgrinberg** [comment]
Perhaps we should split the vscode core and language client bindings into separate libraries `Vscode` and `Vscode_languageclient`? It seems simpler to me to just mirror the upstream structure as much as possible.

File: src/bindings/node/dune [hunk 14]
```diff
@@ -0,0 +1,9 @@
+(library
+ (name node)
+ (modules node)
```
Comment (rgrinberg) [inline src/bindings/node/dune:3]: Now `modules` can be omitted.

File: src/bindings/node/node.ml [hunk 15]
```diff
@@ -0,0 +1,139 @@
+open Interop
+
+module Process = struct
+  val cwd : unit -> string [@@js.global "process.cwd"]
+
+  val platform : string [@@js.global "process.platform"]
+
+  val env : unit -> string Dict.t [@@js.get "process.env"]
```
Comment (rgrinberg) [inline src/bindings/node/node.ml:8]: Is it possible to keep this type completely opaque?

```ocaml
module Env : sig
  type t
  val get : t -> string -> string option
  val set : t -> string -> string -> unit
end
```
and that's it.

**Comment by ulugbekna** [comment]
I would agree with splitting vscode and languageclient; less so about prefixing it with vscode

**Comment by rgrinberg** [comment]
> less so about prefixing it with vscode

We could always unprefix in our import.ml file.

@mnxn I'm curious, why didn't you decide to upstream `Promise_jsoo` to jsoo itself? The promise API exists on every browser, correct? Usually such bindings just live directly in Jsoo.

**Comment by mnxn** [comment]
> @mnxn I'm curious, why didn't you decide to upstream `Promise_jsoo` to jsoo itself? The promise API exists on every browser, correct? Usually such bindings just live directly in Jsoo.

I don't think I'm prepared to contribute such a project to Js_of_ocaml yet. Especially since some of the language/library features used would raise the Js_of_ocaml's minimum OCaml version up from 4.02 (let operators, `List.find_map`, `List.filter_map`). I could try slimming it down and upstreaming in the future.

**Comment by rgrinberg** [comment]
Do consider upstreaming what's possible in the future. Jsoo's own browser bindings are fairly complete because there's a large group of contributors sending PR's.

File: src/Extension.ml [hunk 16]
```diff
@@ -11,37 +11,39 @@ let openTerminalSelectCommandId = "ocaml.open-terminal-select"
...
-      { command = Path.toString bin
-      ; args = Array.of_list args
-      ; options = { env; shell = false }
-      }
+      let command = Path.toString bin in
+      let options =
+        LanguageClient.ExecutableOptions.create ~env ~shell:false ()
+      in
+      LanguageClient.Executable.create ~command ~args ~options ()
 end
```
Comment (ulugbekna) [inline src/Extension.ml:47]: What do you think if we replace these single function modules with just functions? `createServerOptions ()` or `serverOptions ()`

File: src/dune (hunk 11 above)
Comment (mnxn) [reply src/dune:2]: If I change the name to `vscode-ocaml-platform`, dune looks for the `Vscode-ocaml-platform` module as an entry point. That doesn't seem like a very optimal change.

Is there a way to have separate names for the executable and entry-point module?

**Comment by rgrinberg** [comment]
> Is there a way to have separate names for the executable and entry-point module?

Unfortunately not. You can replace the dashes with underscores however. That seems good enough.

**Comment by Khady** [comment]
> Although the size of the generated JS has increased to about 16MB with Js_of_ocaml, the installation vsix package has only increased to 3.5MB from BuckleScript's 1.0MB.

Have you tried to build in release profile to see if the JS generated by jsoo is smaller? From what I see in the makefile you always build with the default profile.

**Comment by mnxn** [comment]
> Have you tried to build in release profile to see if the JS generated by jsoo is smaller? From what I see in the makefile you always build with the default profile.

I can't believe I forgot about this. Much smaller now - only 2.1MB!

Unfortunately It looks like the sourcemap no longer works, even with the `-source-map` flag. I'm not sure if there's another combination of flags which would keep the sourcemap at a small size.

**Comment by rgrinberg** [comment]
Let's compare in both cases to decide? IMO, if we get better stacktraces with larger binaries, then so be it.

**Comment by rgrinberg** [comment]
@mnxn I pushed a few tweaks. This is ready to go.

One small request: can we generate some opam files from the dune-project? We aren't planning to publish this extension on opam, but it would make it easier to install the dependencies using opam.

**Review by rgrinberg — APPROVED** [review]

**Comment by mnxn** [comment]
> One small request: can we generate some opam files from the dune-project?

Sure, should I specify version constraints for any of the libraries?

What values should I use for the `maintainers` and `authors`?

**Comment by rgrinberg** [comment]
> Sure, should I specify version constraints for any of the libraries?

Let's just set the lower bounds to be the versions of dependencies that you've tested.

> What values should I use for the maintainers and authors?

You can list me as the maintainer. For authors, we can put anyone with over 10 commits as listed by `git shortlog -s -n -e`.

**Comment by rgrinberg** [comment]
Btw, you can create a separate package for your bindings as well. Those can actually be published to opam (if you choose to).

**Comment by mnxn** [comment]
> Btw, you can create a separate package for your bindings as well. Those can actually be published to opam (if you choose to).

If you're referring to the VSCode and Node bindings, I don't feel that those bindings are complete enough to justify publishing to opam.

**Comment by ulugbekna** [comment]
Congrats everyone and huge huge thanks to Max and Rudi!

**Comment by yminsky** [comment]
A minor point: if you want to reduce the size of the generated code, you could consider using Base instead of Core_kernel.

**Comment by idkjs** [comment]
Do the docs need to change to reflect this? They still refer to [Bucklescript here](https://github.com/ocamllabs/vscode-ocaml-platform#bucklescript).

**Comment by mnxn** [comment]
> Do the docs need to change to reflect this? They still refer to [Bucklescript here](https://github.com/ocamllabs/vscode-ocaml-platform#bucklescript).

No, those instructions are about how to _use_ the extension for BuckleScript projects. This PR only changed how the extension is built. BuckleScript will be supported with those instructions for the time being.
